node/src/components/block_validator.rs

```diff
@@ -59,114 +68,114 @@ pub(crate) struct BlockValidationState<T> {
+                    return responder.respond((true, block)).ignore();
+                }
 
                 // TODO: Clean this up to use `or_insert_with_key` once
                 // https://github.com/rust-lang/rust/issues/71024 is stabilized.
                 match self.validation_states.entry(block) {
                     Entry::Occupied(mut entry) => {
-                        // The entry already exists. We register ourselves as someone interested in
-                        // the ultimate validation result.
-                        entry.get_mut().responders.push(responder);
+                        // The entry already exists.
+                        if entry.get().missing_deploys.is_empty() {
+                            // Block has already been validated successfully, early return to
+                            // caller.
+                            effects.extend(responder.respond((true, entry.key().clone())).ignore());
+                        } else {
+                            // We register ourselves as someone interested in the ultimate
+                            // validation result.
+                            entry.get_mut().responders.push(responder);
+                        }
                     }
                     Entry::Vacant(entry) => {
                         // Our entry is vacant - create an entry to track the state.
                         let missing_deploys: HashSet<DeployHash> =
                             entry.key().deploys().iter().map(|hash| **hash).collect();
 
+                        let in_flight = &mut self.in_flight;
+                        let chainspec = Arc::clone(&self.chainspec);
+                        let fetch_effects: Effects<Event<T, I>> = block_deploys
+                            .iter()
+                            .flat_map(|deploy_hash| {
+                                let chainspec = Arc::clone(&chainspec);
+                                // For every request, increase the number of in-flight...
+                                in_flight.inc(deploy_hash);
+
+                                // ...then request it.
+                                let deploy_hash = *deploy_hash;
+                                let validate_deploy =
+                                    move |result: FetchResult<Deploy>| match result {
+                                        FetchResult::FromStorage(deploy)
+                                        | FetchResult::FromPeer(deploy, _) => {
+                                            if deploy.header().is_valid(
+                                                &chainspec.genesis.deploy_config,
+                                                block_timestamp,
+                                            ) {
+                                                Event::DeployFound(deploy_hash)
+                                            } else {
+                                                Event::DeployMissing(deploy_hash)
```

Comment:
I am worried that this is swallowing the real reason this block cannot be validated – it turns an _invalid_ deploy into a missing one. It's probably fine for now , I just want to put it out there.

Reply:
True, after all the work of fetching the deploys, we are returning a `bool`. I would like to rethink a few aspects of the `block_validator` at some point - for instance, batching the deploy fetching.

Reply:
> batching the deploy fetching.
Oh yes, definitely that :+1: :100: